### PR TITLE
Feature/async html from next branch

### PR DESCRIPTION
**Pull request by jzeltman**
Issue: Async HTML Stories

## What I did
Updated the HTML Render method to accept asynchronous stories. Discussed with Norbert and found that support had mostly been added already, but just needed to update the render method. I also updated the story snapshots with the new async story.

## How to test
Run tests, or view the async welcome story: http://localhost:9006/?path=/story/welcome--async-welcome

- Is this testable with Jest or Chromatic screenshots? Yes
- Does this need a new example in the kitchen sink apps? Yes. It's been added: http://localhost:9006/?path=/story/welcome--async-welcome
- Does this need an update to the documentation? No

### CODE REVIEW

File: examples/html-kitchen-sink/stories/__snapshots__/welcome.stories.storyshot [hunk 1]
```diff
@@ -99,3 +99,5 @@ exports[`Storyshots Welcome Welcome 1`] = `
 
 </div>
 `;
+
+exports[`Storyshots Welcome Welcome Async 1`] = `[object Promise]`;
```
Comment (ndelangen) [inline examples/html-kitchen-sink/stories/__snapshots__/welcome.stories.storyshot]: This is a bit odd?

Comment (jzeltman) [reply examples/html-kitchen-sink/stories/__snapshots__/welcome.stories.storyshot]: I’m unfamiliar with how to update the storyshots with async stories. Do you know of any examples I could use to update this?

Comment (ndelangen) [reply examples/html-kitchen-sink/stories/__snapshots__/welcome.stories.storyshot]: Looks like we need to do something in storyshots to properly support this.

Comment (ndelangen) [reply examples/html-kitchen-sink/stories/__snapshots__/welcome.stories.storyshot]: We could figure it our together in a pair programming session if you'd like?

**Comment by shilman** [comment]
Related https://github.com/storybookjs/storybook/issues/10009 

cc @tmeasday

**Review by ndelangen — APPROVED** [review]

**Comment by tmeasday** [comment]
@jzeltman thanks for this PR, I am looking at is as part of https://github.com/storybookjs/storybook/issues/10009 and will merge it soon.

One question -- what is the significance of the `.async.js` extension? Does it do anything special? 

It appears like the story you've created, although "async" in the sense of returning a promise, isn't in "async" in the sense of actually not rendering right away. Is that right?

I'm thinking we should add some proper async stories like:

```js
export const WelcomeAsync = async () => {
  await new Promise((r) => setTimeout(r, 5000));
  return welcomeAsync;
};

export const WelcomeAsyncDecorator = () => welcomeAsync;
WelcomeAsyncDecorator.story = {
  decorators: [
    async (s) => {
      await new Promise((r) => setTimeout(r, 5000));
      return s();
    },
  ],
};
```

Note that these work, although highlight that we need to render some interstitial screen during the "waiting" part of an async story (not sure if we need a new event for this). We can do that in a separate PR as part of https://github.com/storybookjs/storybook/issues/10009 though.

**Comment by jzeltman** [comment]
@tmeasday the .async was just to denote that this was an example of an async story. 

We use this format on the storybook/aem version and it works well. The default storybook loading icon displays until the promise resolves. 

It probably also work in the way you’ve outlined above as well though, I haven’t had a chance to test it out. 

I wrote this after discussing it with @ndelangen but it didn’t get merged earlier due to problems with my inexperience with storyshots and storyshots not working easily with async.

**Comment by ndelangen** [comment]
Hey @jzeltman 

Thank you for your time and energy put into this. on the surface this looks great, but @tmeasday and I had to dig a bit deeper, and discovered a pretty big problem.

Storybook has a concept called decorators, these are functions that can wrap the storyFn. They can add a bit of wrapping html or do something with the storyFn.

Decorators are sync functions right now.

It's not feasible for us right now to the change the nature of decorators and make them async.
If users would mix async & sync decorators with sync & async stories, a lot of issues would come up. In particular React doesn't support async functions as components; only sync.

So unfortunately this means we can't really merge this PR.

---

We do recognize the need to do async setup for stories. So we're proposing to introduce a 'loaders' or 'setup' mechanism for stories.

These setup functions would run in parallel BEFORE the storyFn, and their results would be passed into the storyFn as part of the storyContext.

WDYT?

**Pull request closed without merging**